### PR TITLE
Added v2/system

### PR DESCRIPTION
**Pull request by MrMorie**
### Fixes [DPC-1385](https://jira.cms.gov/browse/DPC-1385)

SSAS POST /v2/system will be very similar to SSAS v1 POST /system.  Key differences:

A public key is required
Instead of creating and returning a client id & secret, it should create and return a Client Token.
A new database table tokens will be required.

The tokens table stores data about Client Tokens (macaroons)
There is no updated_at column because tokens are not updatable
The tokens table does not contain any tokens, just a label and a UUID for each token that has been created

### Proposed Changes
- Implement POST v2 Systems route/service
- Create UP/Down migration files for new client tokens table

### Change Details
- Added POST v2/system route
- Created up/down migration files for client_tokens table
- included "updated_at" col since GORM requires it.
- added migration unit tests
- added v2/system tests

### Security Implications

- [ ] new software dependencies

- [ ] security controls or supporting software altered

- [ ] new data stored or transmitted

- [ ] security checklist is completed for this change

- [ ] requires more information or team discussion to evaluate security implications

- [x] no PHI/PII is affected by this change

### Acceptance Validation

Screen Shot
Screen Shot
Screen Shot

### CODE REVIEW

File: ssas/systems.go [hunk 1]
```diff
@@ -72,6 +72,47 @@ type Secret struct {
...
+	Uuid     string `json:"uuid"`
+	System   System `gorm:"foreignkey:SystemID;association_foreignkey:ID"`
+	SystemID uint   `json:"system_id"`
+}
+
+/*
+	SaveClientToken should be provided with a token label and toke uuid, which will
+	be saved to the client tokens table and associated with the current system.
+*/
+func (system *System) SaveClientToken(uuidS string, label string) error {
```
[the remark this answers is not in the document]
Comment (MrMorie) [reply ssas/systems.go:87]: there is a uuid package under that same name that is being used, so there is a name collision. Figured this was easier than giving the package an alias.

File: ssas/service/admin/api_test.go [hunk 2]
```diff
@@ -765,3 +766,139 @@ func contains(list []string, target string) bool {
...
+	assert.NotEmpty(s.T(), result["client_id"])
+	assert.NotEmpty(s.T(), result["client_token"])
+	assert.Empty(s.T(), result["client_secret"])
+	assert.Equal(s.T(), "Test Client", result["client_name"])
+
+	err = ssas.CleanDatabase(group)
+	assert.Nil(s.T(), err)
+}
+
+//V2 Tests Below
```
[the remark this answers is not in the document]
Comment (MrMorie) [reply ssas/service/admin/api_test.go]: fixed

**Comment by MrMorie** [comment]
passed build && package https://bcda-ci.adhocteam.us/view/SSAS/job/SSAS%20-%20Build%20and%20Package/1426/

File: ssas/systems.go [hunk 3]
```diff
@@ -72,6 +72,47 @@ type Secret struct {
...
+	}
+	ClientTokenCreated(Event{Op: "SaveClientToken", TrackingID: uuid.NewRandom().String(), ClientID: system.ClientID})
+	return nil
+}
+
+func (system *System) GetClientTokens(trackingID string) ([]ClientToken, error) {
+	db := GetGORMDbConnection()
+	defer Close(db)
+
+	getEvent := Event{Op: "GetClientTokes", TrackingID: trackingID, Help: "calling from systems.GetClientTokens()"}
```
Comment (dhgreene) [inline ssas/systems.go:108]:
```suggestion
	getEvent := Event{Op: "GetClientToken", TrackingID: trackingID, Help: "calling from systems.GetClientTokens()"}
```

File: ssas/systems.go [hunk 4]
```diff
@@ -444,36 +499,47 @@ func RegisterSystem(clientName string, groupID string, scope string, publicKeyPE
 			return creds, errors.New("internal system error")
 		}
 	}
+	if isV2 {
+		err = system.SaveClientToken(uuid.NewRandom().String(), "Initial Token")
+		if err != nil {
+			regEvent.Help = fmt.Sprintf("could not save client token for clientID %s, groupID %s: %s", clientID, groupID, err.Error())
+			OperationFailed(regEvent)
+			return creds, errors.New("internal system error")
```
Comment (dhgreene) [inline ssas/systems.go:507]: Should we roll back the transaction too?

Comment (MrMorie) [reply ssas/systems.go:508]: yep, fixed.

**Review by dhgreene — APPROVED** [review]
💯

**Pull request merged**